node/node_cross_shard.go

```diff
@@ -120,124 +119,75 @@ func (node *Node) VerifyBlockCrossLinks(block *types.Block) error {
-				lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-				if err != nil {
-					return ctxerror.New("[CrossLinkVerification] no last cross link found 1",
-						"blockHash", block.Hash(),
-						"crossLink", lastLink,
-					).WithCause(err)
-				}
-			}
-		} else {
-			if (*crossLinks)[i-1].Header().ShardID() != crossLink.Header().ShardID() {
-				if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 {
-					lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-					if err != nil {
-						return ctxerror.New("[CrossLinkVerification] no last cross link found 2",
-							"blockHash", block.Hash(),
-							"crossLink", lastLink,
-						).WithCause(err)
-					}
-				}
-			} else {
-				lastLink = &(*crossLinks)[i-1]
-			}
-		}
-
-		if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 { // TODO: verify genesis block
-			err = node.VerifyCrosslinkHeader(lastLink.Header(), crossLink.Header())
-			if err != nil {
-				return ctxerror.New("cannot ValidateNewBlock",
+	for _, crossLink := range *crossLinks {
+		cl, err := node.Blockchain().ReadCrossLink(crossLink.ShardID(), crossLink.BlockNum())
+		if err == nil && cl != nil {
+			if !bytes.Equal(cl.Serialize(), crossLink.Serialize()) {
+				return ctxerror.New("[CrossLinkVerification] Double signed crossLink",
 					"blockHash", block.Hash(),
-					"numTx", len(block.Transactions()),
-				).WithCause(err)
+					"Previous committed crossLink", cl,
+					"crossLink", crossLink,
+				)
 			}
+			continue
+		}
+		if err = node.VerifyCrossLink(crossLink); err != nil {
+			return ctxerror.New("cannot VerifyBlockCrossLinks",
+				"blockHash", block.Hash(),
+				"blockNum", block.Number(),
+				"crossLinkShard", crossLink.ShardID(),
+				"crossLinkBlock", crossLink.BlockNum(),
+				"numTx", len(block.Transactions()),
+			).WithCause(err)
 		}
 	}
 	return nil
 }
 
-// ProcessHeaderMessage verify and process Node/Header message into crosslink when it's valid
-func (node *Node) ProcessHeaderMessage(msgPayload []byte) {
+// ProcessCrossLinkMessage verify and process Node/CrossLink message into crosslink when it's valid
+func (node *Node) ProcessCrossLinkMessage(msgPayload []byte) {
 	if node.NodeConfig.ShardID == 0 {
-
-		var headers []*block.Header
-		err := rlp.DecodeBytes(msgPayload, &headers)
+		var crosslinks []types.CrossLink
+		err := rlp.DecodeBytes(msgPayload, &crosslinks)
 		if err != nil {
 			utils.Logger().Error().
 				Err(err).
-				Msg("[ProcessingHeader] Crosslink Headers Broadcast Unable to Decode")
+				Msg("[ProcessingCrossLink] Crosslink Message Broadcast Unable to Decode")
 			return
 		}
-
-		// Try to reprocess all the pending cross links
-		node.pendingClMutex.Lock()
-		crossLinkHeadersToProcess := node.pendingCrossLinks
-		node.pendingCrossLinks = []*block.Header{}
-		node.pendingClMutex.Unlock()
-
 		firstCrossLinkBlock := core.EpochFirstBlock(node.Blockchain().Config().CrossLinkEpoch)
```

Comment:
If crosslink has Epoch in it, now we can just compare directly the epoch, no need to get the first block anymore as we don't care about the order now (the first block's crosslink doesn't have to be processed first anymore). Also after staking, the logic of EpochFirstBlock will be invalid.